ga4gh/exceptions.py

```diff
@@ -0,0 +1,179 @@
+    server.
+    """
+
+    def getMessage(self):
+        """
+        Returns the message that we map into the GA4GH protocol message.
+        For many exceptions we can simply set the message statically
+        for a given class, and this message will be returned by
+        default. For more elaborate messages that must be contructed
+        at run time, we can override this method and use state held
+        in instance variables in the exception.
+        """
+        return self.message
+
+    @classmethod
+    def getErrorCode(cls):
+        """
+        Returns an integer that can be used to identif this class.
+        """
+        # We use the CRC32 digest of the class name as a unique code.
+        # We follow the recommendation of the Python docs to ensure
+        # that this value is signed 32 bit integer.
+        code = zlib.crc32(cls.__name__) & 0xffffffff
+        return code
+
+    def toProtocolElement(self):
+        """
+        Converts this exception into the GA4GH protocol type so that
+        it can be communicated back to the client.
+        """
+        error = protocol.GAException()
+        error.errorCode = self.getErrorCode()
+        error.message = self.getMessage()
+        return error
+
+#####################################################################
+#
+# Exceptions that occur in the normal operation of the server
+#
+#####################################################################
+
+
+class BadRequestException(ServerException):
+    """
+    A request that we don't like was sent to the server.
+    """
+    httpStatus = 400
+    message = "Bad request"
+
+
+class BadPageSizeException(BadRequestException):
+
+    message = "Request page size invalid"
+
+
+class BadPageTokenException(BadRequestException):
+
+    message = "Request page token invalid"
+
+
+class NotFoundException(ServerException):
+    """
+    The superclass of all exceptions in which some resource was not
+    found.
+    """
+    message = "A resource was not found"
+    httpStatus = 404
+
+
+class PathNotFoundException(NotFoundException):
+
+    message = "The request path was not found"
+
+
+class ObjectNotFoundException(NotFoundException):
+
+    message = "The requested object was not found"
+
+
+class UnsupportedMediaTypeException(ServerException):
+    httpStatus = 415
+    message = "Unsupported media type"
+
+
+class VersionNotSupportedException(NotFoundException):
+
+    message = "API version not supported"
+
+
+class MethodNotAllowedException(ServerException):
+
+    httpStatus = 405
+    message = "Method not allowed"
+
+
+class NotImplementedException(ServerException):
+    """
+    Exception raised when a part of the API has not been implemented.
+    """
+    httpStatus = 501
+
+    def __init__(self, message):
+        self.message = message
+
+
+class RequestValidationFailureException(ServerException):
+    """
+    A validation of the request data failed
+    """
+
```

Comment:
style: no newline between class doc comment and first method (this applies other places in the file as well)